### PR TITLE
Fix right shifting decoder inputs

### LINKED ISSUE #25: Change Adam to AdamW with linear schedule with warmup
Use AdamW from transformers instead of Adam. Use linear schedule with warmup, `transformers.get_linear_schedule_with_warmup`.

### LINKED ISSUE #26: Fix the warmup and total training step calculation for multi-device
Currently we are calculating warmup and total training steps based on data loader length. However, we need to consider the scenario of multi-device training in which case total length will be divided by number of devices.

### LINKED ISSUE #27: Log images every 100 mini-batches
Currently we are logging images every 1000 mini batches during validation and testing. However, we will not have enough images logged in case we have a small dataset or higher number of devices. Either we should change it to 100 mini-batches or take this value as a command line argument.

### PR DESCRIPTION
**Pull request by mahalrs**
Closes #25
Closes #26
Closes #27

- Fixed right shifting decoder inputs that was causing super low training loss.
- Changes embedding size based on decoder inputs
- Fixed tokenizer to ignore special tokens such as decoder start token, BOS, EOS, padding tokens.
- Turn off shuffle for test data loader.

### CODE REVIEW

**Comment by mahalrs** [comment]
@ArjunKrishna3367 @Waleedprw22

File: src/data.py [hunk 1]
```diff
@@ -59,13 +59,9 @@ def get_image_paths(self, dataset_dir, split):
...
 class EncodedVisualNewsDataset(Dataset):
 
-    def __init__(self,
-                 visual_news_dataset_dir,
-                 encoded_data_file,
-                 split,
-                 headlines=False):
-        self.data = self.get_data(visual_news_dataset_dir, encoded_data_file,
-                                  split, headlines)
+    def __init__(self, encoded_dataset, split):
```
Comment (Waleedprw22) [inline src/data.py:62]: so we got the dataset directory out entirely and are sticking with the encoded/processed dataset?

**Review by Waleedprw22 — APPROVED** [review]
certainly a lot of simplications which is good

File: src/data.py (hunk 1 above)
Comment (mahalrs) [reply src/data.py:62]: Yes, this way we are able to remove the logic out of dataset class abut whether to use captions or headlines. It's better to handle that via command line arguments which simplifies the dataset class.

Also, since each encoded file is supposed to only contain captions or headlines, it reduces the time and memory it takes to load the entire data file in memory.

**Comment by ArjunKrishna3367** [comment]
Question: so what does the embedding file look like now?
From my understanding:
headline_tokens/caption_tokens -> input_tokens
headline_attention/caption_attention -> attention_mask
image_tokens -> labels

Did you change the encode_data.py as well? Would be good to add it to this PR as well.

**Comment by mahalrs** [comment]
> Question: so what does the embedding file look like now?
>

What do you mean by embedding file? I made changes to embedding size and input/output sequences. These changes are made in `hparams_base.json` file. Right shifting the decoder inputs in handled in the forward method. The tokenizer is modified to account for these changes.

**Comment by mahalrs** [comment]
> From my understanding: headline_tokens/caption_tokens -> input_tokens headline_attention/caption_attention -> attention_mask image_tokens -> labels
> 
> Did you change the encode_data.py as well? Would be good to add it to this PR as well.

Yes, I made changes to simplify the dataset class. Most of the logic is now handled via command line arguments.

I didn't change `encode_data.py` file. There are additional changes that we need to make to `encode_data.py` file such as split train/val/test data into separate files (#29) when encoding because loading the entire file in memory is super slow and consume too much memory (during init process, it loads data that's not even needed). As we have limited time, I didn't modify `encode_data.py` file but feel free to open a new PR (#29).
